### PR TITLE
Schedule

### PR DESCRIPTION
**Pull request by kathomas921**
This isn't ready to go yet, but I made some changes to creation/modification of schedules.

You now need to enter exactly one task (root task) when you create a schedule.
All non-root tasks need a trigger and parent.
"Save" saves all new tasks and reloads the edit page rather than taking you back to the index (which "Back" does now).
The root task is automatically assigned upon schedule creation.
You can't delete or change the root task (at the moment, there are too many things that can go wrong if you have the ability to change or delete the root task).

Issues:
I am a CSS muggle and have no idea why the "Save" button is the whole width of the page, nor have any idea where to look to fix it.
I cannot figure out how to only allow a user to enter a single task before creating the schedule (ex. having the green button disappear after pressing it once if you haven't created the schedule yet, or eliminating the button for "new" and having a single set of fields show up automatically).

I think that's it for now. Let me know if you have any suggestions for the issues mentioned.

### CODE REVIEW

File: app/models/task.rb [hunk 1]
```diff
@@ -77,9 +77,9 @@ class Task < ActiveRecord::Base
 
   # Ensures that the control pin and the power pin are set to different values or no value.
   def event_and_parent_validation
-    if parent_id.nil? ^ trigger.nil?
-      errors.add(:trigger, ' may be not set unless the Parent Task is specified.') if parent_id.nil?
-      errors.add(:parent_id, ' requires a trigger event.') if trigger.nil?
+    unless schedule.nil? or schedule.root_task_id.nil?
```
Comment (kyleoliveira) [inline app/models/task.rb]: Keep the old validation and make a second validation that handles this case.

Additionally, don't use the `or` method. Just use `||`. Less of an issue here, but `or` has lower operator precedence and is very rarely useful.

File: app/views/schedules/_fields.html.erb [hunk 2]
```diff
@@ -11,6 +11,7 @@
              Task.where(schedule_id: f.object.id).map { |t| ["##{t.id}", t.id] },
              {include_blank: true, selected: (f.object.nil? || f.object.root_task.nil? ? '' : f.object.root_task.id)},
              class: 'form-control' %>
+
```
Comment (kyleoliveira) [inline app/views/schedules/_fields.html.erb]: You'll want to remove these extra spaces in the file if you don't do anything else with it as this progresses. No need to include a whitespace-only change.

File: app/views/schedules/edit.html.erb [hunk 3]
```diff
@@ -5,8 +5,9 @@
   <div class="col-md-8 col-md-offset-2">
     <%= form_for(@schedule) do |f| %>
         <%= render partial: 'fields', locals: {f: f, object: @schedule} %>
-        <%= button_tag 'Save', type: :submit, class: 'btn btn-sm btn-primary' %>
-        <%= link_to 'Cancel', schedules_path, class: 'btn btn-sm btn-primary btn-danger' %>
+
+        <%= button_to 'Save', edit_schedule_path(@schedule), class: 'btn btn-sm btn-primary' %>
```
Comment (kyleoliveira) [inline app/views/schedules/edit.html.erb]: I don't think you need to change these lines for the #edit view. Once the user has been redirected to this view from #create, from then on the user should be able to change the Schedule, save, and carry on as before.

File: app/views/schedules/new.html.erb [hunk 4]
```diff
@@ -4,8 +4,8 @@
 <div class="row">
   <div class="col-md-8 col-md-offset-2">
     <%= form_for(@schedule) do |f| %>
-        <%= render partial: 'fields', locals: {f: f, object: @schedule} %>
-        <%= button_tag 'Add', type: :submit, class: 'btn btn-sm btn-primary' %>
+        <%= render partial: 'fields', locals: {f: f, object: @schedule } %>
+        <%= button_tag 'Add Root Task', type: :submit, class: 'btn btn-sm btn-primary' %>
```
Comment (kyleoliveira) [inline app/views/schedules/new.html.erb]: Since the user is also adding the Name, I don't think you need to change this. The button is referencing that the user is really adding the _Schedule_, not the components of the Schedule.

File: db/seeds.rb [hunk 5]
```diff
@@ -20,4 +20,77 @@
                email: email,
                password:              password,
                password_confirmation: password)
-end
\ No newline at end of file
+end
```
Comment (kyleoliveira) [inline db/seeds.rb:23]: You'll want to rebase off of master before we merge this (but after squashing).

File: app/controllers/schedules_controller.rb [hunk 6]
```diff
@@ -59,27 +60,36 @@ def edit
   end
 
   def create
+    # Check to see if it has a root task before creating a schedule
+    msg = check_root_task(schedule_params[:tasks_attributes])
+    if !msg.nil?
```
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: Use `unless` instead of `if !`.

File: app/controllers/schedules_controller.rb [hunk 7]
```diff
@@ -59,27 +60,36 @@ def edit
...
+      redirect_to new_schedule_path(schedule: params[:schedule]), notice: msg
```
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: Use `flash` with the appropriate level instead of `notice`. You'll see examples in the various controllers' methods.

File: app/models/schedule.rb [hunk 8]
```diff
@@ -18,9 +18,7 @@ class Schedule < ActiveRecord::Base
                                              .pluck(:schedule_id)) }
 
   # == Validators ==
-  validates :root_task, presence: true,
-                        on: :update,
-                        unless: 'tasks.empty? || tasks.none?{ |t| !t.id.nil? }'
+  validates_presence_of :name
```
Comment (kyleoliveira) [inline app/models/schedule.rb:21]: Why did you remove the validation for `:root_task`?

**Comment by kyleoliveira** [comment]
With respect to button sizes, check out the CSS classes provided by [Bootstrap](http://getbootstrap.com/css/) - that's generally how we're handling that stuff.

File: app/views/schedules/edit.html.erb (hunk 3 above)
Comment (kathomas921) [inline app/views/schedules/edit.html.erb]: I'm not sure what you are saying exactly. Before I changed 'Save' to go to edit_schedule_path, it would render the index upon clicking save. I wanted the user to be able to save tasks and stay on the same page rather than saving, being diverted back to index, then having to click into schedule to edit again. I believe that is effectively the only thing I changed. Am I mistaken?

File: app/views/schedules/new.html.erb (hunk 4 above)
Comment (kathomas921) [inline app/views/schedules/new.html.erb]: Since I haven't figured out a way to block the addition of multiple tasks before the root task is saved (and all non-root tasks require a parent task), I phrased it that way to try to remind the user to enter exactly one root task before adding other tasks. It's an easy change back however if you want.

File: app/models/schedule.rb (hunk 8 above)
Comment (kathomas921) [inline app/models/schedule.rb:21]: Ahh, I was playing around with how to assign root task (trying to assign it so I could validate on submit in addition to update). I'll put that back though.

File: app/views/schedules/edit.html.erb (hunk 3 above)
Comment (kyleoliveira) [inline app/views/schedules/edit.html.erb]: Correct, that is the change you've made. What I'm saying is that it is working as intended, so there is no need for that change at the moment. You'll notice that redirecting to the index page after saving on Edit is the standard workflow for most (all?) controllers in the project. Saving and staying on the Edit page is a new use case and if we're changing how we do things we should probably at least write up an activity diagram or some such and brainstorm the right way to go about it.

Case in point, if you'd really like a way to save and stay on the same page, maybe a new button with something like "Save and Continue Editing" would do the trick rather than changing the current button. Or, if we want to get really fancy, you could use Bootstrap to convert the current Save button to a button with a dropdown (see "Add Equipment" on Equipment Profiles) and have it do the default action when you click and offer the "And Continue" option drop down when you click the carat.

File: app/views/schedules/new.html.erb (hunk 4 above)
Comment (kyleoliveira) [inline app/views/schedules/new.html.erb]: Ah, I see. To see how to add a single Task as part of this page, check out what we do to add a ParticleDevice to a Rhizome. You'll probably end up making a new partial (which isn't a problem at all), but you can base it off the existing one and strip the Cocoon-related stuff out.

**Comment by kathomas921** [comment]
I changed back the functionality so 'save' returns to the index.
I made some extra validations to ensure that non root tasks require both parent_id and trigger and that root tasks have neither.

@kyleoliveira Let me know if there is anything you still want changed or implemented before I merge in with master.

File: app/views/schedules/_task_fields.html.erb [hunk 9]
```diff
@@ -19,11 +21,14 @@
                      class: 'form-control' %>
 
         <span class="input-group-btn">
-        <%= link_to_remove_association '',
-                                       f,
-                                       alt: 'Remove Task',
-                                       title: 'Remove Task',
-                                       class: 'btn btn-danger glyphicon glyphicon-minus' %>
+        <% unless !f.object.schedule.nil? and f.object.id == f.object.schedule.root_task_id %>
```
Comment (kyleoliveira) [inline app/views/schedules/_task_fields.html.erb]: Rather than making this view more complicated, you should add fields for the Task under `app/views/schedules/_fields.html.erb`.

File: app/views/schedules/new.html.erb [hunk 10]
```diff
@@ -4,7 +4,7 @@
 <div class="row">
   <div class="col-md-8 col-md-offset-2">
     <%= form_for(@schedule) do |f| %>
-        <%= render partial: 'fields', locals: {f: f, object: @schedule} %>
+        <%= render partial: 'fields', locals: {f: f, object: @schedule } %>
```
Comment (kyleoliveira) [inline app/views/schedules/new.html.erb]: If you drop the space, I think this file won't change... or am I missing something?

File: app/controllers/schedules_controller.rb [hunk 11]
```diff
@@ -128,5 +142,14 @@ def schedule_params
                     tasks_attributes: [:id, :type, :sprout, :equipment, :duration, :trigger,
                                        :target_temperature, :ramp_estimate, :parent_id, :_destroy])
     end
+
+    def check_root_task(task_attributes)
```
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: This should really be a validation in the model rather than a controller method.

Here's what I think you need to do to make that happen:
- Add another validation to Schedule that forces there to only be one Task on #new even outside of the controller. It should be something like:

``` ruby
validates_numericality_of :tasks, { equal_to: 1 }, on: :new
```
- Add an if-statement around lines 19-27 in app/views/schedules/_fields.html.erb so that the button to add another task isn't rendered. When combined with your other guard around the button to remove tasks, I think that'll make it so you have one and only one task on the #new view, which is basically what you're trying to do on line 150.
- Remove the `on: :update` qualifier from the :root_task in the Schedule model. That'll replace the check you're trying to do on line 148 by requiring a root_task all the time.

Once you've done that, you can drop this method and drop the change on lines 63-70.

File: app/controllers/schedules_controller.rb [hunk 12]
```diff
@@ -49,6 +49,7 @@ def new
           'You must save the Schedule first before each Task is assigned an ID.'
     flash.now[:info] = msg
     @schedule = params[:schedule].nil? ? Schedule.new : Schedule.new(schedule_params)
+    @root = Task.new
```
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: If you're not using `@root` in the view, I think you can move the creation of this root task into the Schedule #initialize method.

Doing that would also replace Schedule#assign_root and (hopefully) make it so you don't have to save twice on create.

File: db/seeds.rb [hunk 13]
```diff
@@ -93,4 +93,4 @@
 BeerRecipe.create!(
   name: "Test Recipe",
   schedule_id: sched.id,
-  )
+  )
```
Comment (kyleoliveira) [inline db/seeds.rb:96]: You should be able to revert the change for this file :wink:

**Comment by kyleoliveira** [comment]
@kathomas921 Updated comments. Let me know if you've got questions and/or grab me on Hangouts if you need more help.

File: app/controllers/schedules_controller.rb [hunk 14]
```diff
@@ -59,18 +59,26 @@ def edit
   end
 
   def create
-    @schedule = Schedule.new(schedule_params)
+    # Check to see if it has a root task before creating a schedule
+    msg = check_root_task(schedule_params[:tasks_attributes])
+    unless msg.nil?
+      flash[:notice] = msg
```
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: Change this to `flash[:danger]`. The symbol used for flash messages corresponds to [CSS in Bootstrap](http://getbootstrap.com/css/#helper-classes) - see other instances of `flash` for more examples.

File: app/controllers/schedules_controller.rb (hunk 11 above)
Comment (kyleoliveira) [inline app/controllers/schedules_controller.rb]: We determined it's `validates_length_of` instead.

File: app/views/schedules/_fields.html.erb [hunk 15]
```diff
@@ -16,13 +16,15 @@
...
     <div class="add-task-control">
-    <%= link_to_add_association '',
-                                f,
-                                :tasks,
-                                'data-association-insertion-method': :before,
-                                'data-association-insertion-traversal': :parent,
-                                alt: 'Add Task',
-                                title: 'Add Task',
-                                class: 'btn btn-sm btn-success glyphicon glyphicon-plus task' %>
+      <%= link_to_add_association '',
```
Comment (kyleoliveira) [inline app/views/schedules/_fields.html.erb:19]: Is this just an indentation change?

File: app/views/schedules/_root_task_fields.html.erb [hunk 16]
```diff
@@ -0,0 +1,46 @@
+<div class="nested-fields form-inline">
+  <div class="form-group">
+    <div class="input-group">
+        <% if f.object.id.nil? %>
```
Comment (kyleoliveira) [inline app/views/schedules/_root_task_fields.html.erb]: Since this partial is only meant to be used when we're creating the Schedule, the Task's id will always be nil. You can remove the other case.

File: app/views/schedules/_root_task_fields.html.erb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+                     class: 'form-control' %>
+        <span class="input-group-addon"></span>
+        <%= f.label :sprout, class: 'sr-only' %>
+        <%= f.select :sprout,
+                     EquipmentProfile.equipment_profile_options,
+                     {include_blank: true, selected: (f.object.sprout.nil?  ? '' : f.object.sprout_name)},
+                     class: 'form-control' %>
+
+        <span class="input-group-btn">
+        <% unless !f.object.schedule.nil? and f.object.id == f.object.schedule.root_task_id %>
```
Comment (kyleoliveira) [inline app/views/schedules/_root_task_fields.html.erb]: Similarly, since this is only for the creation action and we're limiting this to one Task, you should be able to remove everything in this `unless` block.

File: app/views/schedules/_task_fields.html.erb [hunk 18]
```diff
@@ -19,11 +19,14 @@
...
```
Comment (kyleoliveira) [inline app/views/schedules/_task_fields.html.erb]: We don't have to do this now, right?

If we do... change `and` to `&&`.

Comment (kathomas921) [inline app/views/schedules/_task_fields.html.erb]: That is just a clause so that the remove symbol shows up next to each task except the root task (so you can't accidentally delete the root). I can (and subsequently have) gotten rid of the first part of the unless clause checking if the schedule exists (it always will when task_fields is called).

**Comment by kyleoliveira** [comment]
If you revert the changes to the files that are only changing indentation or line endings, and squash this into a single commit, I think this will be ready to go. I haven't tested it out yet, but I'll take your word on it.
